src/ZODB/Connection.py

```diff
@@ -757,7 +758,11 @@ def afterCompletion(self, transaction):
         #    finalizing previous ones without calling begin.  We pass
         #    False to avoid possiblyt expensive sync calls to not
         #    penalize well-behaved applications that call begin.
-        self.newTransaction(transaction, False)
+        if hasattr(self._storage, 'afterCompletion'):
+            self._storage.afterCompletion()
```

Comment:
WRT "rolling back twice", I wondered about that a little.  It's conceivable that this could do work if you started a new transaction right after ending one, but avoiding that seems like the responsibility of the storage.  Ultimately, this is like an event that the storage can do with what it will.